### PR TITLE
Add When I capture HTTP $httpMethod request with URL pattern `$urlPat…

### LINKED ISSUE #394: [Proxy] New step is needed to capture body of the request and status code
**Is your feature request related to a problem? Please describe.**
Step 'When I capture HTTP httpMethod request with URL pattern `urlPattern` and save URL query to scopes variable `variableName`' captures only URL, but not the body of the request.

There is a need also to capture:
- response status
- body of the request

### PR DESCRIPTION
**Pull request by ngrudnitsky**
…tern` and save request data to $scopes variable `$variableName`
Closes #394

### CODE REVIEW

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 1]
```diff
@@ -145,6 +148,62 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+     * <li><b>SCENARIO</b> - the variable will be available only within the scenario,
+     * <li><b>STORY</b> - the variable will be available within the whole story,
+     * <li><b>NEXT_BATCHES</b> - the variable will be available starting from next batch
+     * </ul>
+     * @param variableName A variable name
+     * @throws IOException If any error happens during operation
+     */
+    @When("I capture HTTP $httpMethod request with URL pattern `$urlPattern` and save request data to $scopes "
+            + "variable `$variableName`")
+    public void captureRequestAndSaveRequestData(HttpMethod httpMethod, String urlPattern, Set<VariableScope> scopes,
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:178]: Please add integration test

**Comment by valfirst** [comment]
It would be good to add integration tests as well

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 2]
```diff
@@ -145,6 +148,62 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+            String variableName) throws IOException
+    {
+        List<HarEntry> harEntries = checkNumberOfRequests(httpMethod, urlPattern, ComparisonRule.EQUAL_TO, 1);
+        if (harEntries.size() == 1)
+        {
+            HarEntry harEntry = harEntries.get(0);
+            HarPostData postData = harEntry.getRequest().getPostData();
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:185]: `harEntry.getRequest()` cache it please

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 3]
```diff
@@ -145,6 +148,62 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+     * @param variableName A variable name
+     * @throws IOException If any error happens during operation
+     */
+    @When("I capture HTTP $httpMethod request with URL pattern `$urlPattern` and save request data to $scopes "
+            + "variable `$variableName`")
+    public void captureRequestAndSaveRequestData(HttpMethod httpMethod, String urlPattern, Set<VariableScope> scopes,
+            String variableName) throws IOException
+    {
+        List<HarEntry> harEntries = checkNumberOfRequests(httpMethod, urlPattern, ComparisonRule.EQUAL_TO, 1);
+        if (harEntries.size() == 1)
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:182]: don't you mind to refactor the method into something like this:
```
            requestDataMap.put("queryMap", queryMap(..));
            requestDataMap.put("requestBodyParamsMap", requestBodyParamsMap(..));
            requestDataMap.put("requestBodyMap", requestBodyMap(..));
            requestDataMap.put("responseStatusMap", responseStatusMap(..));
```

File: vividus-plugin-web-app/src/test/java/org/vividus/proxy/steps/ProxyStepsTests.java [hunk 4]
```diff
@@ -137,35 +144,96 @@ void checkHarEntryExistenceWithHttpMethodAndUrlPatternNoCalls() throws IOExcepti
...
+        spy.captureRequestAndSaveRequestData(HTTP_METHOD, URL_PATTERN, variableScopes, VARIABLE_NAME);
+        verify(bddVariableContext).putVariable(eq(variableScopes), eq(VARIABLE_NAME), argThat(value ->
+        {
+            Map<String, Map<String, String>> map = (Map<String, Map<String, String>>) value;
+            Map<String, String> queryMap = map.get("queryMap");
+            Map<String, String> requestBodyMap = map.get("requestBodyMap");
+            Map<String, String> requestBodyParamsMap = map.get("requestBodyParamsMap");
+            Map<String, String> responseStatusMap = map.get("responseStatusMap");
+            String status = "status";
+            return map.size() == 4
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/test/java/org/vividus/proxy/steps/ProxyStepsTests.java:220]: use Assertions.assertAll(...)
In the current implementation, I have to debug and compare manually to understand what is not matching

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 5]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+        {
+            HarEntry harEntry = harEntries.get(0);
+            HarRequest request = harEntry.getRequest();
+            HarPostData postData = request.getPostData();
+            Map<String, Map<String, String>> requestDataMap = new HashMap<>();
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:185]: - construct Map using Map.of(...) method
- do not use collection name as part of variable name

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 6]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+            requestDataMap.put("queryMap", queryMap(request));
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:186]: queryMap - not clear, consider something like getQueryParameters, the same for all the methods below

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 7]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+            requestDataMap.put("requestBodyParamsMap", requestBodyParamsMap(postData));
+            requestDataMap.put("requestBodyMap", requestBodyMap(postData));
+            requestDataMap.put("responseStatusMap", responseStatusMap(harEntry));
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:189]: variable names for end users should be simplified and made clearer

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 8]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+            bddVariableContext.putVariable(scopes, variableName, requestDataMap);
+        }
+    }
+
+    private Map<String, String> responseStatusMap(HarEntry harEntry)
+    {
+        int statusCode = harEntry.getResponse().getStatus();
+        return Map.of("status", String.valueOf(statusCode));
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:197]: why not to put status as plain value? we can not get more than one status so there is no reason to put it as a map

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 9]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+    }
+
+    private Map<String, String> requestBodyMap(HarPostData postData)
+    {
+        Map<String, String> requestBodyMap = new HashMap<>();
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:202]: use map.of(...)

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 10]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+        requestBodyMap.put("mimeType", postData.getMimeType());
+        requestBodyMap.put("text", postData.getText());
+        requestBodyMap.put("comment", postData.getComment());
+        return requestBodyMap;
+    }
+
+    private Map<String, String> requestBodyParamsMap(HarPostData postData)
+    {
+        return postData.getParams()
+                        .stream()
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:212]: not aligned

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java [hunk 11]
```diff
@@ -138,13 +142,84 @@ public void captureRequestAndSaveURLQuery(HttpMethod httpMethod, String urlPatte
...
+                        .collect(Collectors.toMap(HarPostDataParam::getName, HarPostDataParam::getValue));
+    }
+
+    private Map<String, String> queryMap(HarRequest request)
+    {
+        return request.getQueryString()
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:218]: not aligned

File: vividus-plugin-web-app/src/test/java/org/vividus/proxy/steps/ProxyStepsTests.java [hunk 12]
```diff
@@ -137,35 +145,89 @@ void checkHarEntryExistenceWithHttpMethodAndUrlPatternNoCalls() throws IOExcepti
...
+        when(harRequest.getQueryString())
+                .thenReturn(List.of(firstPair, secondPair));
+        String mimeType = "mimeType";
+        String text = "text";
+        String comment = "comment";
+        HarPostData postData = new HarPostData();
+        postData.setMimeType(mimeType);
+        postData.setText(text);
+        postData.setComment(comment);
+        HarPostDataParam postDataParam1 = new HarPostDataParam();
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/test/java/org/vividus/proxy/steps/ProxyStepsTests.java:202]: constructions of HarPostDataParam, HarQueryParam objects can be moved into separate method, please request similar logic

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java (hunk 8 above)
Comment (ngrudnitsky) [reply vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:197]: We use Map<String, Map<String, String>> to save all needed information to variable, so we can't put status as plain value into this object

Comment (uarlouski) [reply vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:197]: Map<String, Object> ?

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java (hunk 9 above)
Comment (ngrudnitsky) [reply vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:202]: I can't use Map.of in this plaсe, cause it returns ImmutableCollections and HarPostData.getComment can return null when it's not presented

File: vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java (hunk 8 above)
Comment (ngrudnitsky) [reply vividus-plugin-web-app/src/main/java/org/vividus/proxy/steps/ProxySteps.java:197]: If we use Map<String, Object> then we can't use varibles like ${requestData.requestBody.text} to check if it's saved correctly and we can't compare whole map, cause it's order is not guaranteed

**Pull request merged**